scalyr_agent/tests/copying_manager_test.py

```diff
@@ -784,6 +821,186 @@ def test_pipelined_requests_with_retry_error(self):
+        controller = self.__create_test_instance()
+        previous_root_dir = os.path.dirname(self.__test_log_file)
+
+        self.__append_log_lines('First line', 'Second line')
+        (request, responder_callback) = controller.wait_for_rpc()
+        lines = self.__extract_lines(request)
+
+        self.assertEquals(2, len(lines))
+        self.assertEquals('First line', lines[0])
+        self.assertEquals('Second line', lines[1])
+
+        controller.stop()
+
+        fp = open(os.path.join(self._config.agent_data_path, 'active-checkpoints.json'), 'r')
+        data = json_lib.parse(fp.read())
+        fp.close()
+
+        data['time'] += 100
+
+        fp = open(os.path.join(self._config.agent_data_path, 'active-checkpoints.json'), 'w')
+        fp.write(json_lib.serialize(data))
+        fp.close()
+
+        self.__append_log_lines('Third line', 'Fourth line')
+
+        controller = self.__create_test_instance(root_dir=previous_root_dir, auto_start=False)
+        self._manager.start_manager(dict(fake_client=True))
+
+        (request, responder_callback) = controller.wait_for_rpc()
+        lines = self.__extract_lines(request)
+        self.assertEquals(2, len(lines))
+        self.assertEquals('Third line', lines[0])
+        self.assertEquals('Fourth line', lines[1])
+
+    def test_start_without_active_checkpoint(self):
+        controller = self.__create_test_instance()
+        previous_root_dir = os.path.dirname(self.__test_log_file)
+
+        self.__append_log_lines('First line', 'Second line')
+        (request, responder_callback) = controller.wait_for_rpc()
+        lines = self.__extract_lines(request)
+
+        self.assertEquals(2, len(lines))
+        self.assertEquals('First line', lines[0])
+        self.assertEquals('Second line', lines[1])
+        controller.stop()
+
+        self.__append_log_lines('Third line', 'Fourth line')
+
+        os.remove(os.path.join(self._config.agent_data_path, 'active-checkpoints.json'))
+
+        controller = self.__create_test_instance(root_dir=previous_root_dir, auto_start=False)
+        self._manager.start_manager(dict(fake_client=True))
+        (request, responder_callback) = controller.wait_for_rpc()
+        lines = self.__extract_lines(request)
+        self.assertEquals(2, len(lines))
+        self.assertEquals('Third line', lines[0])
+        self.assertEquals('Fourth line', lines[1])
+
+    def test_stale_request(self):
+        controller = self.__create_test_instance()
+        self.__append_log_lines('First line', 'Second line')
+        (request, responder_callback) = controller.wait_for_rpc()
+
+        lines = self.__extract_lines(request)
+        self.assertEquals(2, len(lines))
+        self.assertEquals('First line', lines[0])
+        self.assertEquals('Second line', lines[1])
+
+        from scalyr_agent import copying_manager
+
+        # backup original 'time' module
+        orig_time = copying_manager.time
+
+        class _time_mock(object):
```

Comment:
Wrap this in a try/finally block so that if there is an exception, you will always restore `copying_manager.time` to its original value.

Reply:
Please correct me if I missed something, but which kinds of exceptions we can face here? During copying manager runtime or during declaration and mocking?  In both cases there are only one standard lib time.time() call and few assignments. Do you think they really can fail?

Reply:
I agree with you that the code as it stands can't fail.  However, what happens if someone comes along and adds a new line that can throw an exception?  They might not notice that they need to do a try/finally to protect against it.  I think it's generally a good habit and future proofs the code if you use the try/finally pattern when you acquire resources or need to be sure to undo side effects.

Reply:
That's a very good point. Thank you.